File: mkdocs/overrides/devbook/network-currency/query-block-rewards.mjs
import { PublicKey } from 'symbol-sdk';
import { Address, Network } from 'symbol-sdk/symbol';

const fmt = (v) => (v / 1e6).toLocaleString(
	'en-US', { minimumFractionDigits: 6 });

const NODE_URL = process.env.NODE_URL ||
	'https://reference.symboltest.net:3001';
console.log(`Using node ${NODE_URL}`);

const BLOCK_HEIGHT = process.env.BLOCK_HEIGHT || '3222290';

// Get the block header [>step-1]
const block = await (await fetch(
	`${NODE_URL}/blocks/${BLOCK_HEIGHT}`)).json();
const signer = Network.TESTNET.publicKeyToAddress(
	new PublicKey(block.block.signerPublicKey));
const beneficiary = block.block.beneficiaryAddress;
console.log(`Block height: ${BLOCK_HEIGHT}`);
console.log(`Signer: ${signer}`);
const beneficiaryB32 = Address.fromDecodedAddressHexString(
	beneficiary);
console.log(`Beneficiary: ${beneficiaryB32}`);
// [<step-1]
// Get the network sink address [>step-2]
const properties = await (await fetch(
	`${NODE_URL}/network/properties`)).json();
const sinkB32 = properties.chain.harvestNetworkFeeSinkAddress;
const sink = Array.from(new Address(sinkB32).bytes)
	.map(b => b.toString(16).padStart(2, '0')).join('').toUpperCase();
console.log(`Network sink: ${sinkB32}`);
// [<step-2]
// Get the inflation reward at this height [>step-3]
const inflation = await (await fetch(
	`${NODE_URL}/network/inflation/at/${BLOCK_HEIGHT}`)).json();
const reward = parseInt(inflation.rewardAmount, 10);
console.log(`Inflation reward: ${fmt(reward)} XYM`);
// [<step-3]
// Get harvest fee receipts for this block [>step-4]
const receipts = await (await fetch(
	`${NODE_URL}/statements/transaction`
	+ `?height=${BLOCK_HEIGHT}&receiptType=8515`)).json();

// Label and display the reward distribution
let total = 0;
console.log('\nReward distribution:');
for (const item of receipts.data) {
	for (const r of item.statement.receipts) {
		if (r.type !== 8515) continue;
		const amount = parseInt(r.amount, 10);
		total += amount;
		let label;
		if (r.targetAddress === sink) {
			label = 'Network sink (5%)';
		} else if (r.targetAddress === beneficiary) {
			label = 'Beneficiary (25%)';
		} else {
			label = 'Harvester';
			const harvester = Address.fromDecodedAddressHexString(
				r.targetAddress);
			console.log(`  ${label}: ${fmt(amount)} XYM`);
			console.log(`  Harvester: ${harvester}`);
			continue;
		}
		console.log(`  ${label}: ${fmt(amount)} XYM`);
	}
}
// [<step-4]
// Summary [>step-5]
const fees = total - reward;
console.log('\nSummary:');
console.log(`  Total block reward: ${fmt(total)} XYM`);
console.log(`  Inflation: ${fmt(reward)} XYM`);
console.log(`  Transaction fees: ${fmt(fees)} XYM`); // [<step-5]

File: mkdocs/overrides/devbook/network-currency/query-block-rewards.py
import json
import os
import urllib.request

from symbolchain.CryptoTypes import PublicKey
from symbolchain.symbol.Network import Address, Network

NODE_URL = os.getenv(
	'NODE_URL', 'https://reference.symboltest.net:3001')
print(f'Using node {NODE_URL}')

BLOCK_HEIGHT = os.getenv('BLOCK_HEIGHT', '3222290')

try:
	# Get the block header
	with urllib.request.urlopen(
			f'{NODE_URL}/blocks/{BLOCK_HEIGHT}') as response:
		block = json.loads(response.read())
	signer = Network.TESTNET.public_key_to_address(
		PublicKey(block['block']['signerPublicKey']))
	beneficiary = block['block']['beneficiaryAddress']
	print(f'Block height: {BLOCK_HEIGHT}')
	print(f'Signer: {signer}')
	beneficiary_b32 = Address.from_decoded_address_hex_string(
		beneficiary)
	print(f'Beneficiary: {beneficiary_b32}')

	# Get the network sink address
	with urllib.request.urlopen(
			f'{NODE_URL}/network/properties') as response:
		properties = json.loads(response.read())
	sink_b32 = properties['chain']['harvestNetworkFeeSinkAddress']
	sink = Address(sink_b32).bytes.hex().upper()
	print(f'Network sink: {sink_b32}')

	# Get the inflation reward at this height
	with urllib.request.urlopen(
			f'{NODE_URL}/network/inflation'
			f'/at/{BLOCK_HEIGHT}') as response:
		inflation = json.loads(response.read())
	reward = int(inflation['rewardAmount'])
	print(f'Inflation reward: {reward / 1e6:,.6f} XYM')

	# Get harvest fee receipts for this block
	with urllib.request.urlopen(
			f'{NODE_URL}/statements/transaction'
			f'?height={BLOCK_HEIGHT}'
			f'&receiptType=8515') as response:
		receipts = json.loads(response.read())

	# Label and display the reward distribution
	total = 0
	print('\nReward distribution:')
	for item in receipts['data']:
		for r in item['statement']['receipts']:
			if r['type'] != 8515:
				continue
			amount = int(r['amount'])
			total += amount
			target = r['targetAddress']
			if target == sink:
				label = 'Network sink (5%)'
			elif target == beneficiary:
				label = 'Beneficiary (25%)'
			else:
				label = 'Harvester'
				print(f'  {label}: {amount / 1e6:,.6f} XYM')
				harvester = Address.from_decoded_address_hex_string(
					target)
				print(f'  Harvester: {harvester}')
				continue
			print(f'  {label}: {amount / 1e6:,.6f} XYM')

	# Summary
	fees = total - reward
	print('\nSummary:')
	print(f'  Total block reward: {total / 1e6:,.6f} XYM')
	print(f'  Inflation: {reward / 1e6:,.6f} XYM')
	print(f'  Transaction fees: {fees / 1e6:,.6f} XYM')

except Exception as error:
	print(error)
